plugins/dell-dock2/fu-dell-dock2-ec.c

```diff
@@ -0,0 +1,972 @@
+
+	/* erase */
+	if (!fu_dell_dock2_hid_erase_bank(device, 0xff, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* write to device */
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autoptr(FuChunk) chk = NULL;
+
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+
+		if (!fu_dell_dock2_ec_hid_write(device, fu_chunk_get_bytes(chk), error))
+			return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	if (!fu_dell_dock2_hid_raise_mcu_clock(device, FALSE, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	g_debug("ec firmware written successfully");
+
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_query_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	/* dock data */
+	if (!fu_dell_dock2_ec_dock_data_cmd(device, error))
+		return FALSE;
+
+	/* dock info */
+	if (!fu_dell_dock2_ec_dock_info_cmd(device, error))
+		return FALSE;
+
+	/* set internal dock sku, must after dock info */
+	if (!fu_dell_dock2_ec_set_dock_sku(device, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_reload(FuDevice *device, GError **error)
+{
+	/* if query looks bad, wait a few seconds and retry */
+	if (!fu_device_retry_full(device, fu_dell_dock2_ec_query_cb, 10, 2000, NULL, error)) {
+		g_prefix_error(error, "failed to query dock ec: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_setup(FuDevice *device, GError **error)
+{
+	guint32 ec_version = 0;
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_dell_dock2_ec_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* get dock type */
+	if (!fu_dell_dock2_ec_dock_type_cmd(device, error))
+		return FALSE;
+
+	/* if query looks bad, wait a few seconds and retry */
+	if (!fu_device_retry_full(device, fu_dell_dock2_ec_query_cb, 10, 2000, NULL, error)) {
+		g_prefix_error(error, "failed to query dock ec: ");
+		return FALSE;
+	}
+
+	/* setup version */
+	ec_version = GUINT32_FROM_BE(fu_dell_dock2_ec_get_ec_version(device));
+	fu_device_set_version_raw(device, ec_version);
+
+	/* create the subcomponents */
+	if (!fu_dell_dock2_ec_probe_subcomponents(device, error))
+		return FALSE;
+
+	g_debug("dell-dock2-ec->setup done successfully");
+	return TRUE;
+}
+
+static gchar *
+fu_dell_dock2_ec_convert_version(FuDevice *device, guint64 version_raw)
+{
+	return fu_hex_version_from_uint32(version_raw, fu_device_get_version_format(device));
+}
+
+static gboolean
+fu_dell_dock2_ec_open(FuDevice *device, GError **error)
+{
+	/* FuUdevDevice->open */
+	return FU_DEVICE_CLASS(fu_dell_dock2_ec_parent_class)->open(device, error);
+}
+
+static void
+fu_dell_dock2_ec_finalize(GObject *object)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(object);
+	g_free(self->dock_data);
+	g_free(self->dock_info);
+	g_free(self->raw_versions);
+	g_free(self->ec_minimum_version);
```

Comment:
I don't see this variable being used, think it can go.

Reply:
Fixed, thanks.